### PR TITLE
CollectDataToIORank: move to separate compile unit

### PR DESCRIPTION
**Pull request by akva2**
make it a template for grid types. this allows using
explicit template instantation and compile this code
only once per grid type, instead of once per simulator object.

this might be a more controversial change, opinions wanted.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this opm-grid=531 please

**Comment by akva2** [comment]
jenkins build this opm-grid=531 please

**Comment by akva2** [comment]
jenkins build this please

**Comment by blattms** [comment]
I generally like this but am wondering:
- Shouldn't we make it possible for downstream users to use other grids?
- Can we deduce the grid view type?

**Comment by akva2** [comment]
1) if this is a priority we need to reshuffle the code a little again.
2) possibly, it's a separate entity in the typetag so i kept it that way.

File: ebos/collecttoiorank.cc [hunk 1]
```diff
@@ -0,0 +1,902 @@
...
+
+
+
+template class CollectDataToIORank<Dune::CpGrid,
+                                   Dune::CpGrid,
+                                   Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid>>>;
+template class CollectDataToIORank<Dune::PolyhedralGrid<3,3,double>,
+                                   Dune::PolyhedralGrid<3,3,double>,
+                                   Dune::GridView<Dune::PolyhedralGridViewTraits<3,3,double,Dune::PartitionIteratorType(4)>>>;
+
```
Comment (totto82) [inline ebos/collecttoiorank.cc:901]: I use the CollectDataToIORank in my adaptive gridding prototype simulator based on AluGrid.  Note the `needsReordering` I haven't read the code in details, but is it sufficient to add a AluGrid version here to make it compile and run with AluGrid?

Comment (akva2) [reply ebos/collecttoiorank.cc:901]: yes.

Comment (totto82) [reply ebos/collecttoiorank.cc:901]: >but in no way does it change the implementation
Good to know. It isn't always easy to tell from the `diff ` when it involves renaming.

**Comment by akva2** [comment]
so, should we make it extendible or not? i would like to point out that this was not the case before either, as none of the headers in ebos/ are installed.

**Comment by blattms** [comment]
If it was not the case before, we should not put in extra effort now.

**Comment by blattms** [comment]
@totto82 should check the branch with his code, though.

**Comment by totto82** [comment]
I fine with this. From my point of view this can be merged.

**Pull request merged**